File: SortVision/public/code/insertion/python/insertionSort.py
def insertion_sort(arr: list[int]) -> list[int]:
    # Traverse from the second element to the end of the list
    for i in range(1, len(arr)):
        key = arr[i]  # The current element to be inserted
        j = i - 1
        # Move elements of arr[0..i-1], that are greater than key, to one position ahead
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        # Insert the key at the correct position
        arr[j + 1] = key
    return arr

File: SortVision/public/code/insertion/typescript/insertionSort.ts
function insertionSort(arr: number[]): number[] {
    // Make a copy of the input array to avoid modifying the original
    const array = [...arr];
    
    // Start from the second element (index 1)
    // as a single element is already sorted
    for (let i = 1; i < array.length; i++) {
        // Store the current element to be inserted
        const current = array[i];
        let j = i - 1;
        
        // Move elements that are greater than current
        // to one position ahead of their current position
        while (j >= 0 && array[j] > current) {
            array[j + 1] = array[j];
            j--;
        }
        
        // Place the current element in its correct position
        array[j + 1] = current;
    }
    
    return array;
}

